framework/web/UrlRule.php

```diff
@@ -367,8 +367,15 @@ public function createUrl($manager, $route, $params)
                 continue;
             }
             if (!isset($params[$name])) {
-                return false;
-            } elseif (strcmp($params[$name], $value) === 0) { // strcmp will do string conversion automatically
+                // allow omit empty optional params
+                // @see https://github.com/yiisoft/yii2/issues/10970
+                if (in_array($name, $this->placeholders) && strcmp($value, '') === 0) {
```

Comment:
@rob006 I get some puzzled, I'll try to explain what I want to do:

I have rules, let's say like this
`
[
           'pattern' => '<lang>/page/<id:\d+>',
           'route' => 'site/view',
           'host' => 'http://yii2.app',
           'defaults' => ['lang' => '']
]
`

First, when I tried to make url with `Url::to(['site/view', 'id' => 123]);`
the output was /site/view?id=123, when I applied your patch (from here), the url was like `http://yii2.app/page/123` and that's ok, but when I set some data to lang (let's say `'lang' => 'en'`) in defaults, my url returned to /site/view?id=123 type, but I expected something like http://yii2.dev/en/page/123

The fact is that value in defaults is global and can be set based on some conditions (`'defaults' => ['lang' => $someValue]`), and I don't want to set this value in all my urls when generate them